File: android/src/main/kotlin/com/loopit/minis/imgedit/ImageEditPluginRouter.kt
package com.loopit.minis.imgedit

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Single shared MethodChannel + EventChannel for the image editor. Every
 * Dart call carries a `viewId` argument; the router dispatches it to the
 * matching [ImageEditEngine].
 */
object ImageEditPluginRouter {
    private const val METHOD_NAME = "loopit/minis/imgedit"
    private const val EVENT_NAME = "loopit/minis/imgedit/state"

    private val engines = HashMap<Int, ImageEditEngine>()
    private var methodChannel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var stateSink: EventChannel.EventSink? = null

    fun attach(messenger: BinaryMessenger) {
        if (methodChannel != null) return
        val mc = MethodChannel(messenger, METHOD_NAME)
        mc.setMethodCallHandler { call, result -> dispatch(call, result) }
        methodChannel = mc
        val ec = EventChannel(messenger, EVENT_NAME)
        ec.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
                stateSink = events
            }
            override fun onCancel(arguments: Any?) {
                stateSink = null
            }
        })
        eventChannel = ec
    }

    fun detach() {
        methodChannel?.setMethodCallHandler(null)
        methodChannel = null
        eventChannel?.setStreamHandler(null)
        eventChannel = null
        stateSink = null
        engines.values.toList().forEach { it.detach() }
        engines.clear()
    }

    fun register(engine: ImageEditEngine) {
        engines[engine.viewId] = engine
    }

    fun unregister(viewId: Int) {
        engines.remove(viewId)
    }

    fun notifyState(kind: String, payload: Map<String, Any?>) {
        val sink = stateSink ?: return
        val event = HashMap<String, Any?>(payload.size + 1)
        event["kind"] = kind
        event.putAll(payload)
        sink.success(event)
    }

    fun notifyRenderProgress(viewId: Int, pct: Double) =
        notifyState("renderProgress", mapOf("viewId" to viewId, "pct" to pct))

    fun notifyError(viewId: Int, code: String, message: String) =
        notifyState("error", mapOf("viewId" to viewId, "code" to code, "message" to message))

    fun notifyMemoryPressure(viewId: Int?, levelMb: Int) =
        notifyState("memoryPressure", mapOf("viewId" to viewId, "levelMb" to levelMb))

    private fun dispatch(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "listFilters" -> {
                result.success(mapOf("filters" to AssetCatalog.filters()))
                return
            }
            "listStickerPacks" -> {
                result.success(mapOf("packs" to AssetCatalog.stickerPacks()))
                return
            }
            "listFonts" -> {
                result.success(mapOf("fonts" to AssetCatalog.fonts()))
                return
            }
        }
        val args = call.arguments as? Map<*, *>
        val viewId = (args?.get("viewId") as? Number)?.toInt()
        val engine = viewId?.let { engines[it] } ?: engines.values.firstOrNull()
        if (engine == null) {
            result.success(emptyMap<String, Any?>())
            return
        }
        engine.handle(call, result)
    }
}

File: ios/Classes/ImgEdit/ImageEditPluginRouter.swift
import Flutter
import Foundation
import UIKit

/// Single shared MethodChannel + EventChannel for the image editor.
/// Every Dart call carries a `viewId`; the router dispatches it to the
/// matching `ImageEditEngine`. Mirrors `ImageEditPluginRouter.kt`.
public final class ImageEditPluginRouter: NSObject {
    public static let shared = ImageEditPluginRouter()

    public static let methodChannelName = "loopit/minis/imgedit"
    public static let eventChannelName = "loopit/minis/imgedit/state"
    public static let platformViewType = "loopit/minis/imgedit/canvas"

    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?
    private var engines: [Int64: ImageEditEngine] = [:]

    public func attach(messenger: FlutterBinaryMessenger) {
        guard methodChannel == nil else { return }
        let mc = FlutterMethodChannel(
            name: ImageEditPluginRouter.methodChannelName,
            binaryMessenger: messenger
        )
        mc.setMethodCallHandler { [weak self] call, result in
            self?.handle(call: call, result: result)
        }
        methodChannel = mc

        let ec = FlutterEventChannel(
            name: ImageEditPluginRouter.eventChannelName,
            binaryMessenger: messenger
        )
        ec.setStreamHandler(EventStream { [weak self] sink in
            self?.eventSink = sink
        })
        eventChannel = ec
    }

    public func detach() {
        methodChannel?.setMethodCallHandler(nil)
        methodChannel = nil
        eventChannel?.setStreamHandler(nil)
        eventChannel = nil
        eventSink = nil
        for engine in engines.values { engine.detach() }
        engines.removeAll()
    }

    func register(_ engine: ImageEditEngine) {
        engines[engine.viewId] = engine
    }

    func unregister(viewId: Int64) {
        engines.removeValue(forKey: viewId)
    }

    func notifyState(kind: String, payload: [String: Any]) {
        guard let sink = eventSink else { return }
        var event = payload
        event["kind"] = kind
        sink(event)
    }

    public func notifyRenderProgress(viewId: Int64, pct: Double) {
        notifyState(kind: "renderProgress", payload: ["viewId": viewId, "pct": pct])
    }

    public func notifyError(viewId: Int64, code: String, message: String) {
        notifyState(kind: "error",
                    payload: ["viewId": viewId, "code": code, "message": message])
    }

    public func notifyMemoryPressure(viewId: Int64?, levelMb: Int) {
        var payload: [String: Any] = ["levelMb": levelMb]
        if let id = viewId { payload["viewId"] = id }
        notifyState(kind: "memoryPressure", payload: payload)
    }

    private var memoryObserver: NSObjectProtocol?

    public func observeMemoryWarnings() {
        let center = NotificationCenter.default
        memoryObserver = center.addObserver(
            forName: UIApplication.didReceiveMemoryWarningNotification,
            object: nil, queue: .main
        ) { [weak self] _ in
            self?.notifyMemoryPressure(viewId: nil, levelMb: 0)
        }
    }

    public func stopObservingMemoryWarnings() {
        if let obs = memoryObserver {
            NotificationCenter.default.removeObserver(obs)
            memoryObserver = nil
        }
    }

    private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "listFilters":
            result(["filters": AssetCatalog.filters()])
            return
        case "listStickerPacks":
            result(["packs": AssetCatalog.stickerPacks()])
            return
        case "listFonts":
            result(["fonts": AssetCatalog.fonts()])
            return
        default:
            break
        }

        let args = (call.arguments as? [String: Any]) ?? [:]
        let viewId: Int64
        if let raw = args["viewId"] as? NSNumber {
            viewId = raw.int64Value
        } else {
            viewId = engines.keys.first ?? -1
        }
        guard let engine = engines[viewId] else {
            result([:])
            return
        }
        engine.handle(call: call, result: result)
    }
}

private final class EventStream: NSObject, FlutterStreamHandler {
    let onListen: (FlutterEventSink?) -> Void

    init(onListen: @escaping (FlutterEventSink?) -> Void) {
        self.onListen = onListen
    }

    func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        onListen(events)
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        onListen(nil)
        return nil
    }
}
